roles/init/tasks/postgres.yml

```diff
@@ -247,16 +247,37 @@
   with_dict: >-
     {{ postgres_tablespaces|default({}) }}
 
+# Determine the upstream primary for this host, that is, the primary at
+# the root of the cascading replication setup. This may be the instance
+# itself, if it's a primary.
+
+- name: Record name of the upstream primary
+  set_fact:
+    upstream_primary: "{{ inventory_hostname|upstream_root(hostvars) }}"
+  when: >
+    'postgres' in role
+
+- block:
+  - name: Prevent repmgr is installed on wrong BRD nodes when creating a replica
+    assert:
+      msg: >-
+        The upstream must be a subscriber-only node
```

Comment:
Also see the equivalent assert in roles/postgres/replica/final. Perhaps we should move that here and use that message (or something else that explains the problem in more detail).

Reply:
the assert in roles/postgres/replica/final already has a reasonable explanation, so if you move assert here, then that should take care of the explanation bit.

Aside, I was wondering whether it would make sense to introduce the assert in validate-topology.yml and fail much earlier?

Reply:
Sounds reasonable. I bet this might not be the only case. So I think we could start thinking where to move them once we gather them.
However, it sounds something beyond the scope of this PR.

Reply:
The change to move the assert? IMHO, we should handle that as part of this PR to avoid duplication of checks at different places.

Reply:
Have a look at https://github.com/EnterpriseDB/tpa/pull/35/files#diff-50f10ac800fc23b3ae37d01aea48c407946a64c0e6370d600d2509b3129f6af8 because it's not duplicated any more.

What I'm saying is your idea to "fail sooner" is good. I thought you were talking about something like one big playbook with tons of those checks that could run at the begging. Sorry.

Reply:
no, my idea to fail sooner was to deal with this problem in validate-topology.yml because imho it's also a somewhat topological problem i.e. we don't want to consume a config.yml that specifies replicas of non-so nodes in a PGD cluster. Anyway, this works too.